Docs/rewards (#119)
archway-network/archway-docs: docs/rewards/managing-rewards.md

---
sidebar_position: 2
---

# Managing Rewards 

> This guide will go through how to rewards are managed on the Archway Network by explaining the important fields and operations needed to track and receive rewards. 

Reviewer: ```suggestion
This guide will go through how rewards are managed on the Archway Network by explaining the important fields and operations needed to track and receive rewards.
```

Author: Re: Minimum Consensus Fee - Understand the importance here, I tried to keep this content solely focused on rewards. Obviously, these are connected, do you have an idea where you would like to see it specifically added? 

Re: Top Level - When I started I was told these pages are owned by the design team so I have not edited them. The config file for adding this would require me to change the docusaurus config files included in the gitignore. Please confirm if I should be the one making these changes.

## Contract Metadata 

> Contract Metadata is the metadata that instructs the Archway Network who ones the contract and where rewards should be sent. The metadata has 3 specific parameters: 

Reviewer 2: "3" should be "three."

> Contract Metadata is the metadata that instructs the Archway Network who ones the contract and where rewards should be sent. The metadata has 3 specific parameters: 

Reviewer: ```suggestion
Contract Metadata is the metadata that instructs the Archway Network who owns the contract and where rewards should be sent. The metadata has three specific parameters: 
```

> `contract_address` - The CosmWasm address of the contract 

Reviewer: ```suggestion
`contract_address` - The address of the deployed contract 
```

`owner_address` - The address of the contract's owner. This field can be both an account or contract address.

> `rewards_address` - The account address that will receive the contract's rewards. Tokens are sent to this address after a [withdrawal function](#withdrawing-awards) is executed. If not set, the contract will not receive rewards. 

Reviewer: ```suggestion
`rewards_address` - The account address that will receive the contract's rewards. This field can be both an account or contract address. Tokens are sent to this address after executing a [withdrawal function](#withdrawing-awards). If not set, the contract will not receive rewards. 
```

> `rewards_address` - The account address that will receive the contract's rewards. This field can be both an account or contract address. Tokens are sent to this address after executing a [withdrawal function](#withdrawing-awards) is executed. If this field is not set, the contract will not receive rewards. 

Reviewer 3: ```suggestion
`rewards_address` - The account address that will receive the contract's rewards. This field can be both an account or contract address. Tokens are sent to this address after executing a [withdrawal function](#withdrawing-awards). If this field is not set, the contract will not receive rewards. 
```

**All above addresses are bech32-encoded.**

### Setting The Contract Metadata 

> The contract metadata must be set up by using the `MsgSetContractMetadata` transaction which must be signed by a contract admin. 

Reviewer: Please, prefer using an active voice:

```suggestion
The contract metadata must be set up using the `MsgSetContractMetadata` transaction, which a contract admin must sign. 
```

```go  
message ContractMetadata { 
  string contract_address =  "archway14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sy85n2u";
  string owner_address = "archway12reqvcenxgv5s7z96pkytzajtl4lf2epyfman2";
  string rewards_address = "archway12reqvcenxgv5s7z96pkytzajtl4lf2epyfman2";
}

> ``` 

Reviewer: For clarity, it's better to show this information in a JSON object instead.

Reviewer: The comment was not only about changing the code block language from `go` to `json` but also about the content. This is not a valid JSON :)

> ``` 

Reviewer: This is not a valid JSON

Author: Good call, I pulled this from the spec which is using proto3 syntax so that is why I used Go before. Can you send me exactly what this is in JSON?

Reviewer: ```suggestion
```json  
{ 
  "contract_address": "archway14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sy85n2u",
  "owner_address": "archway12reqvcenxgv5s7z96pkytzajtl4lf2epyfman2",
  "rewards_address": "archway12reqvcenxgv5s7z96pkytzajtl4lf2epyfman2"
}
``` 
```

Author: Thanks!

### Updating the Contract Metadata 

> The contract metadata of a contract can be updated by the `owner_address`. This could either be a contract admin or the address of the contract itself. Updates are completed by using the `MsgSetContractMetadata` transaction. 
> 
> If the `owner_address` and `contract_address` are the same, updates to the metadata can be done by using the WASM bindings. 

Reviewer: ```suggestion
The `owner_address` is the only address that can update the metadata of a contract. This could either be an account or a smart contract's address. Updates are completed by using the `MsgSetContractMetadata` transaction.

If the `owner_address` is a smart contract, updates to the metadata can only be performed using the WASM bindings.
```

## Reward Storage

> The rewards owed to a certain contract are stored on the blockchain. To receive these rewards, a [withdrawal operation](#withdrawing-awards) must be executed. Developers can also track these rewards by using the [RewardsRecord object](#tracking-reward-amounts). 

Reviewer: ```suggestion
The rewards owed to a specific contract are stored as separate records on the blockchain. To receive these rewards, a [withdrawal operation](#withdrawing-awards) must be executed. Developers can also track these rewards using the [RewardsRecord object](#tracking-reward-amounts). 
```

## Tracking Reward Amounts 

> The RewardsRecord object is used to track the rewards that have been calculated for a specific `rewards_address`. 

Reviewer: ```suggestion
The `RewardsRecord` object is used to track the rewards that have been calculated for a specific `rewards_address`. 
```

### Example Request - RewardsRecord 
```bash
archwayd q rewards rewards-records "${rewards_address}" --node "${node_url}" --output json | jq .
```
### Example Response 

```json
{
  "id": "1",
  "rewards_address": "archway12reqvcenxgv5s7z96pkytzajtl4lf2epyfman2",
  "rewards": [
    {
      "denom": "uarch",
      "amount": "10000"
    }
  ],
  "calculated_height": 100,
  "calculated_time": {
    "seconds": 1660591975,
    "nanos": 0
  }
}
```

> `id` - Each Reward Record has a unique ID. This can be used to only withdraw certain reward records. 
> 
> `rewards_address` - The account address that will receive the contract's rewards. Tokens are sent to this address after a [withdrawal function](#withdrawing-awards) is executed. If not set, the contract will not receive rewards. 
> 
> `denom`- The token type/denmoination. uarch is the default token of the Archway Network.  
> 
> `amount`- The calculated reward amount. 
> 
> `calculated_height` - The block height of the rewards when the rewards were calculated. 
> 
> `calculated_time` - The block time of when the rewards were calculated. 

Reviewer: ```suggestion
`id` - Each Reward Record has a unique ID. This can be used to withdraw only certain reward records.

`rewards_address` - The account address that will receive the contract's rewards. Tokens are sent to this address after executing a [withdrawal function](#withdrawing-awards). If not set, the contract will not receive rewards. 

`denom`- The token type/denomination. `uarch` is the default token of the Archway Network. The testnets use other denominators, like `uconst` and `utitus`.

`amount`- The calculated reward amount. 

`calculated_height` - The block height of the rewards when the rewards were calculated. 

`calculated_time` - The block time of when the rewards were calculated. 
```

## Withdrawing Awards 

> After rewards are calculated, a withdrawal operation is needed to be executed before they are sent to the `rewards_address`. 
> 
> The withdrawal operation is completed by using the `MsgWithdrawRewards` message.       

Reviewer: ```suggestion
After rewards are calculated, a withdrawal operation has to be executed before they are sent to the `rewards_address`. 

The withdrawal operation is completed by using the `MsgWithdrawRewards` message.

Only the address informed in the `rewards_address` can claim the rewards.
```

### Example Request

```bash
archwayd tx rewards withdraw-rewards --records-limit 100000 --from "${rewards_address}" --node "${node_url}" --chain-id "${chain_id}" 
```

>  ``record_limit`` - Is the maximum number of the Reward Records that be withdrawn. If the limit is 0, the default limit is used in the withdrawal. 
> 
>  ``record_ids`` - If you would like to withdraw specific Reward Records, an array of those ids can be included.
> 
>  ``reward_address`` - The reward address that has been set in the [Contract Metadata](#contract-metadata). 
> 
>  ``node`` - Node to be connected to. 
> 
>  ``chain-id`` - The Archway Chain ID that the contract is on. A list of the available chain IDs can be [found here](../overview/network.md/). 

Reviewer: ```suggestion
 ``record_limit`` - Is the maximum number of the Reward Records that be withdrawn. If the limit is 0, the default limit is used in the withdrawal. 

 ``record_ids`` - If you want to withdraw specific Reward Records, an array of those ids can be included.

 ``reward_address`` - The reward address set in the [Contract Metadata](#contract-metadata).

 ``node`` - Endpoint of the node to be connected to.

 ``chain-id`` - The Archway Chain ID that the contract is on. A list of the available chain IDs can be [found here](../overview/network.md/). 
```

>  ### Response 

Reviewer 2: There's s leading space here; it won't show up properly in most markdown interpreters.

Reviewer 2: Well, VS code seems to be okay with it... so I guess if it's been tested, this should be okay. I won't make this a blocker.

> If the Withdraw Rewards operations was successful, the response will show the number of reward records processed and the total amount of rewards that have been transferred to the `rewards_address`. 

Reviewer: ```suggestion
If the Withdraw Rewards operation is successful, the response will show the number of reward records processed and the total amount of rewards that have been transferred to the `rewards_address`. 
```

> If the Withdraw Rewards operations iss successful, the response will show the number of reward records processed and the total amount of rewards that have been transferred to the `rewards_address`. 

Reviewer 3: "iss successful" should be " is successful"

## Querying Outstanding Rewards

To track what rewards have not been withdrawn and sent to the `rewards_adress` yet, use the below command: 

```bash
archwayd q rewards outstanding-rewards "${rewards_address}" --node "${node_url}" --output json | jq .
``` 

> For information on the available queries and parameters connected to rewards, go to the [Rewards Reference](./rewards-reference.md) page. 

Reviewer: ```suggestion
Go to the [Rewards Reference](./rewards-reference.md) page for information on the available queries and parameters connected to rewards. 
```
